### PR TITLE
ARROW-17792: [C++] Use lambda capture move construction

### PR DESCRIPTION
**Pull request by pitrou**
Simplify some code using functors for the purpose of capturing a moved value.

I did not modify all such places, only where readability would seem to improve.

### CODE REVIEW

**Comment by pitrou** [comment]
@kou Have you already seen this error? https://github.com/apache/arrow/actions/runs/3096909253/jobs/5012993282#step:10:202
```
[58/106] Generating symbol file arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols
FAILED: arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols 
"C:/msys64/ucrt64/bin/meson" "--internal" "symbolextractor" "D:/a/arrow/arrow/build/c_glib" arrow-flight-glib/libarrow-flight-glib-1000.dll "arrow-flight-glib/libarrow-flight-glib.dll.a" arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols
```

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/dataset/dataset_writer.cc [hunk 1]
```diff
@@ -174,19 +174,9 @@ class DatasetWriterFileQueue {
...
-          : self(self), batch(std::move(batch)) {}
-      Result<Future<>> operator()(util::AsyncTaskScheduler*) {
-        return self->WriteNext(std::move(batch));
-      }
-      DatasetWriterFileQueue* self;
-      std::shared_ptr<RecordBatch> batch;
-    };
-    scheduler_->AddTask(
-        ::arrow::internal::make_unique<WriteTask>(this, std::move(batch)));
+    scheduler_->AddSimpleTask([self = this, batch = std::move(batch)]() {
```
Comment (westonpace) [inline cpp/src/arrow/dataset/dataset_writer.cc:177]: Do we want to specify all arguments in this way or just the ones that we need to capture specially?  E.g.  Is `[self = this, batch = std::move(batch)]` preferred over `[this, batch = std::move(batch)]`?

**Review by westonpace — COMMENTED** [review]
Thanks for putting this together.  This has been obnoxious to workaround and I am enjoying this more concise syntax a lot.

File: cpp/src/arrow/dataset/dataset_writer.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/dataset_writer.cc:177]: That's a good question. I think either way is ok.

**Comment by kou** [comment]
> @kou Have you already seen this error? https://github.com/apache/arrow/actions/runs/3096909253/jobs/5012993282#step:10:202
> 
> ```
> [58/106] Generating symbol file arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols
> FAILED: arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols 
> "C:/msys64/ucrt64/bin/meson" "--internal" "symbolextractor" "D:/a/arrow/arrow/build/c_glib" arrow-flight-glib/libarrow-flight-glib-1000.dll "arrow-flight-glib/libarrow-flight-glib.dll.a" arrow-flight-glib/libarrow-flight-glib-1000.dll.p/libarrow-flight-glib-1000.dll.symbols
> ```

Sorry. I missed this.
I have seen the error but it was disappeared without any changes... So I don't know why this is happened yet.

**Pull request merged**